README.md
# liquid_bar_chart


## Animated Bar Chart
![Liquid Bar Chart](https://github.com/mi6ock/liquid_bar_chart/blob/master/readme_files/liquid_bar_chart.gif)

### Description



### Getting started
You should ensure that you add the dependency in your flutter project.

```yaml
dependencies:
  liquid_bar_chart: ^0.0.1
```


## Example

```dart
Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    
    final List<String> week = ["Sun", "Mon", "Thu", "Wed", "Thu", "Fri", "Sat"];
    final List<String> tooltips = [
      "Good",
      "Bad",
      "So So",
      "A little nice",
      "Pretty good",
      "Very good",
      "Super nice"
    ];
    
    Random rnd = Random();
    for (int i=0; i< 7; i++) {
      barDataList.add(
        BarData(
          value: rnd.nextDouble() * 10,
          maxValue: 10,
          barBackColor: Colors.blue[100],
          barColor: Colors.blue,
          title: week[i],
          toolTip: tooltips[i],
        ),
      );
    }
    
    return Scaffold(
      appBar: AppBar(
        title: Text("Chart Example"),
      ),
      body: Center(
        child: SizedBox(
          height: size.height / 2,
          width: size.width,
          child: Card(
            margin: const EdgeInsets.all(18),
            shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.circular(16),
            ),
            color: Colors.blueGrey[100],
            child: LBChart(
              barDataList: barDataList,
              padding: EdgeInsets.all(8),
              title: "Chart Title",
            ),
          ),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => _changeData(),
      ),
    );
}

```

CHANGELOG.md
## [0.0.1] - Oct 15, 2019

test/liquid_bar_chart_test.dart
import 'package:flutter_test/flutter_test.dart';

void main() {
  test('chart check', () {
    expect(3, 3);
  });
}


lib/liquid_bar_chart/chart.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:liquid_bar_chart/liquid_bar_chart/bar_data.dart';
import 'package:liquid_bar_chart/liquid_bar_chart/clipper.dart';

class LBChart extends StatelessWidget {
  List<Widget> widgetList = [];

  final List<BarData> barDataList;
  final backGroundColor;
  final title;
  final description;
  final textStyle;
  final padding;
  double waveHeight;

  LBChart({
    Key key,
    @required this.barDataList,
    this.padding,
    this.backGroundColor,
    this.title,
    this.description,
    this.textStyle,
    this.waveHeight,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    if (barDataList.length < 2) {
      throw Exception("2 or more data is required");
    }

    return Padding(
      padding: padding ?? EdgeInsets.zero,
      child: Center(
        child: LayoutBuilder(builder: (context, snapshot) {
          double height = snapshot.maxHeight;
          double width = snapshot.maxWidth;

          for (var barData in barDataList) {
            widgetList.add(
              WaveBar(
                barData: barData,
                height: height,
                width: width / barDataList.length,
                waveHeight: waveHeight ?? 10.0,
              ),
            );
          }

          return Column(
            children: <Widget>[
              SizedBox(
                height: title != null ? height * 0.15 : 0,
                child: Align(
                  alignment: Alignment.center,
                  child: Text(
                    title ?? "",
                    style: textStyle ?? Theme.of(context).textTheme.title,
                  ),
                ),
              ),
              SizedBox(
                height: title != null ? height * 0.85 : height,
                child: Row(
                  mainAxisAlignment: MainAxisAlignment.spaceBetween,
                  children: widgetList,
                ),
              ),
            ],
          );
        }),
      ),
    );
  }
}

class WaveBar extends StatefulWidget {
  final BarData barData;
  final height;
  final width;
  final double waveHeight;

  const WaveBar({
    Key key,
    @required this.barData,
    this.height,
    this.width,
    this.waveHeight,
  }) : super(key: key);

  @override
  _WaveBarState createState() => _WaveBarState();
}

class _WaveBarState extends State<WaveBar> with SingleTickerProviderStateMixin {
  AnimationController _animationController;

  @override
  void initState() {
    super.initState();
    if (widget.barData.value > widget.barData.maxValue)
      throw Exception("Do not enter a value greater than the maximum value");

    _animationController = AnimationController(
      vsync: this,
      duration: Duration(seconds: 4),
    );
    _animationController.addListener(() => setState(() {}));
    _animationController.repeat();
  }

  @override
  void dispose() {
    _animationController.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;
    double barHeight = widget.barData.title != null
        ? widget.height * 0.7
        : widget.height * 0.85;

    double ratio = widget.barData.value / widget.barData.maxValue;
    double maxCutPoint = (barHeight - widget.width / 2) / barHeight;
    double minCutPoint = widget.width / 2 / barHeight;

    return SizedBox(
      height: widget.height,
      width: widget.width,
      child: Column(
        children: <Widget>[
          Tooltip(
            message: widget.barData.toolTip,
            child: SizedBox(
              height: barHeight,
              width: widget.width,
              child: Card(
                elevation: 4,
                margin: EdgeInsets.symmetric(horizontal: widget.width / 10),
                color: widget.barData.barBackColor,
                shape: RoundedRectangleBorder(
                  borderRadius: BorderRadius.circular(size.width),
                ),
                child: Stack(
                  children: [
                    Align(
                      alignment: Alignment.bottomCenter,
                      child: FractionallySizedBox(
                        widthFactor: 1,
                        // if heightFactor 1, layout broken
                        heightFactor: min(ratio, maxCutPoint) < minCutPoint
                            ? 0
                            : min(ratio, maxCutPoint),
                        child: AnimatedBuilder(
                          animation: CurvedAnimation(
                            parent: _animationController,
                            curve: Curves.linear,
                          ),
                          builder: (context, child) => ClipPath(
                            child: Card(
                              elevation: 0,
                              margin: EdgeInsets.zero,
                              shape: RoundedRectangleBorder(
                                borderRadius: BorderRadius.vertical(
                                  bottom: Radius.circular(size.width),
                                ),
                              ),
                              color: widget.barData.barColor,
                            ),
                            clipper: WaveClipper(
                              animationValue: _animationController.value,
                              ratio: widget.barData.value /
                                  widget.barData.maxValue,
                              waveHeight: widget.waveHeight,
                            ),
                          ),
                        ),
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ),
          SizedBox(
            height: widget.barData.title != null ? widget.height * 0.15 : 0,
            child: Align(
              alignment: Alignment.center,
              child: widget.barData.title != null
                  ? Text(
                      widget.barData.title,
                      style: widget.barData.titleTextStyle ??
                          Theme.of(context).textTheme.title,
                    )
                  : Text(""),
            ),
          ),
        ],
      ),
    );
  }
}


lib/liquid_bar_chart/clipper.dart
import 'dart:math' as math;

import 'package:flutter/material.dart';

class WaveClipper extends CustomClipper<Path> {
  final double animationValue;
  final double ratio;
  final double waveHeight;

  WaveClipper({
    @required this.animationValue,
    @required this.ratio,
    @required this.waveHeight,
  });

  @override
  Path getClip(Size size) {
    Path path = Path()
      ..addPolygon(_wave(size), false)
      ..lineTo(size.width, size.height)
      ..lineTo(0.0, size.height)
      ..close();
    return path;
  }

  List<Offset> _wave(Size size) {
    double dim = 180;
    if (size.width < 50) {
      dim = 90;
    }

    final waves = <Offset>[];
    for (int i = -10; i <= size.width.toInt() + 10; i++) {
      final y = math.sin(
                (animationValue * 360 - i * 2) % 360 * (math.pi / dim),
              ) *
              waveHeight +
          waveHeight;
      waves.add(Offset(i.toDouble(), y));
    }
    return waves;
  }

  @override
  bool shouldReclip(WaveClipper oldClipper) =>
      animationValue != oldClipper.animationValue;
}


lib/liquid_bar_chart/bar_data.dart
import 'package:flutter/material.dart';

class BarData {
  /// chart bar value
  final double value;

  /// chart bar max value (value <= maxValue)
  final double maxValue;

  final Color barBackColor;
  final Color barColor;

  final String title;
  final TextStyle titleTextStyle;
  final String toolTip;

  ///default 4
  final elevation;

  const BarData({
    Key key,
    @required this.value,
    @required this.maxValue,
    @required this.barBackColor,
    @required this.barColor,
    @required this.title,
    @required this.toolTip,
    this.elevation,
    this.titleTextStyle,
  });

  @override
  String toString() {
    return 'BarData{value: $value, maxValue: $maxValue, '
        'barBackColor: $barBackColor, barColor: $barColor, '
        'title: $title, toolTip: $toolTip}';
  }
}


example/lib/main.dart
import 'dart:math';

import 'package:flutter/material.dart';
import 'package:liquid_bar_chart/liquid_bar_chart.dart';

void main() {
  runApp(
    MaterialApp(
      home: ChartExample(),
    ),
  );
}

class ChartExample extends StatefulWidget {
  @override
  _ChartExampleState createState() => _ChartExampleState();
}

class _ChartExampleState extends State<ChartExample> {
  List<BarData> barDataList = [];
  final List<String> week = ["Sun", "Mon", "Thu", "Wed", "Thu", "Fri", "Sat"];
  final List<String> tooltips = [
    "Good",
    "Bad",
    "So So",
    "A little nice",
    "Pretty good",
    "Very good",
    "Super nice"
  ];

  @override
  void initState() {
    super.initState();
    _createData();
  }

  @override
  Widget build(BuildContext context) {
    Size size = MediaQuery.of(context).size;

    return Scaffold(
      appBar: AppBar(
        title: Text("Chart Example"),
      ),
      body: Center(
        child: SizedBox(
          height: size.height / 2,
          width: size.width,
          child: Card(
            margin: const EdgeInsets.all(18),
            shape: RoundedRectangleBorder(
              borderRadius: BorderRadius.circular(16),
            ),
            color: Colors.blueGrey[100],
            child: LBChart(
              barDataList: barDataList,
              padding: EdgeInsets.all(8),
              title: "Chart Title",
            ),
          ),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () => _createData(),
      ),
    );
  }

  void _createData() {
    Random rnd = Random();
    barDataList = [];
    for (int i = 0; i < 7; i++) {
      barDataList.add(
        BarData(
          value: rnd.nextDouble() * 10,
          maxValue: 10,
          barBackColor: Colors.blue[100],
          barColor: Colors.blue,
          title: week[i],
          toolTip: tooltips[i],
        ),
      );
      setState(() {});
    }
  }
}
